curp/src/server/raw_curp/mod.rs

```diff
@@ -219,21 +209,30 @@ impl<C: 'static + Command> RawCurp<C> {
             .ctx
             .others
             .iter()
-            .map(|id| {
+            .filter_map(|id| {
+                #[allow(clippy::indexing_slicing)]
+                // check if heartbeat has been optimized out
+                if self.ctx.hb_opt[id]
+                    .compare_exchange(true, false, Ordering::Acquire, Ordering::Relaxed)
+                    .is_ok()
+                {
+                    return None;
+                }
                 let next_index = lst_r.get_next_index(id);
                 let (prev_log_index, prev_log_term) = log_r.get_prev_entry_info(next_index);
                 debug!("{} send heartbeat to {}", self.id(), id);
-                (
+                let req = (
                     id.clone(),
                     AppendEntries {
                         term,
                         leader_id: self.id().clone(),
                         prev_log_index,
                         prev_log_term,
                         leader_commit: log_r.commit_index,
-                        entries: vec![],
+                        entries: log_r.get_from(next_index).unwrap_or_default().to_vec(),
```

Comment:
If we can calibrate in the heartbeat, do we still need a manually calibrate? Can we merge them into one function?

Reply:
Yes, we can merge them. Actually, I am planning to merge send_log_to_follower, heartbeat, and calibrate all together since all these functionalities use append_entries. And I think after merging them together, batching would be easier. Perhaps for another PR?

Reply:
I can merge them in my new pr